File: checker-go/internal/infrastructure/container/container.go
package container

import (
	"github.com/hidessh99/sub-cf-vpn/checker-go/internal/config"
	"github.com/hidessh99/sub-cf-vpn/checker-go/internal/delivery/http/handler"
	impl "github.com/hidessh99/sub-cf-vpn/checker-go/internal/repository"
	"github.com/hidessh99/sub-cf-vpn/checker-go/internal/infrastructure/logger"
	"github.com/hidessh99/sub-cf-vpn/checker-go/internal/pkg/geoip"
	"github.com/hidessh99/sub-cf-vpn/checker-go/internal/usecase"
	"gorm.io/gorm"
)

type Container struct {
	DB               *gorm.DB
	Config           *config.AppConfig
	Logger           *logger.LogrusLogger
	ProxyUseCase     usecase.ProxyUseCase
	AuthHandler      *handler.AuthHandler
	ProxyHandler     *handler.ProxyHandler
	DomainHandler    *handler.DomainHandler
	BugHandler       *handler.BugHandler
	DashboardHandler *handler.DashboardHandler
	SystemHandler    *handler.SystemHandler
}

func Wire(db *gorm.DB, cfg *config.AppConfig, log *logger.LogrusLogger) *Container {
	// 1. Repositories
	adminRepo := impl.NewAdminRepository(db)
	proxyRepo := impl.NewProxyRepository(db)
	domainRepo := impl.NewDomainRepository(db)
	bugRepo := impl.NewBugRepository(db)

	// 2. Services
	geoIPService := geoip.NewGeoIPService()

	// 3. Use Cases
	authUseCase := usecase.NewAuthUseCase(adminRepo, cfg, log)
	proxyUseCase := usecase.NewProxyUseCase(proxyRepo, geoIPService, cfg, log)
	domainUseCase := usecase.NewDomainUseCase(domainRepo, log)
	bugUseCase := usecase.NewBugUseCase(bugRepo, log)
	dashboardUseCase := usecase.NewDashboardUseCase(proxyRepo, domainRepo, bugRepo, log)

	// 4. Handlers
	authHandler := handler.NewAuthHandler(authUseCase, log)
	proxyHandler := handler.NewProxyHandler(proxyUseCase, log)
	domainHandler := handler.NewDomainHandler(domainUseCase, log)
	bugHandler := handler.NewBugHandler(bugUseCase, log)
	dashboardHandler := handler.NewDashboardHandler(dashboardUseCase, log)
	systemHandler := handler.NewSystemHandler(db, log)

	return &Container{
		DB:               db,
		Config:           cfg,
		Logger:           log,
		ProxyUseCase:     proxyUseCase,
		AuthHandler:      authHandler,
		ProxyHandler:     proxyHandler,
		DomainHandler:    domainHandler,
		BugHandler:       bugHandler,
		DashboardHandler: dashboardHandler,
		SystemHandler:    systemHandler,
	}
}

File: checker/src/container.ts
import { db } from "../database/database";
import { AdminRepository } from "./repositories/AdminRepository";
import { ProxyRepository } from "./repositories/ProxyRepository";
import { DomainRepository } from "./repositories/DomainRepository";
import { BugRepository } from "./repositories/BugRepository";

import { AuthUseCase } from "./usecases/AuthUseCase";
import { ProxyUseCase } from "./usecases/ProxyUseCase";
import { DomainUseCase } from "./usecases/DomainUseCase";
import { BugUseCase } from "./usecases/BugUseCase";
import { DashboardUseCase } from "./usecases/DashboardUseCase";

import { AuthController } from "./controllers/AuthController";
import { ProxyController } from "./controllers/ProxyController";
import { DomainController } from "./controllers/DomainController";
import { BugController } from "./controllers/BugController";
import { DashboardController } from "./controllers/DashboardController";
import { SystemController } from "./controllers/SystemController";

import { GeoIPService } from "./services/GeoIPService";

// 1. Repositories
export const adminRepo = new AdminRepository(db);
export const proxyRepo = new ProxyRepository(db);
export const domainRepo = new DomainRepository(db);
export const bugRepo = new BugRepository(db);

// 2. Services
export const geoIPService = new GeoIPService();

// 3. Use Cases
export const authUseCase = new AuthUseCase(adminRepo);
export const proxyUseCase = new ProxyUseCase(proxyRepo, geoIPService);
export const domainUseCase = new DomainUseCase(domainRepo);
export const bugUseCase = new BugUseCase(bugRepo);
export const dashboardUseCase = new DashboardUseCase(proxyRepo, domainRepo, bugRepo);

// 4. Controllers
export const authController = new AuthController(authUseCase);
export const proxyController = new ProxyController(proxyUseCase);
export const domainController = new DomainController(domainUseCase);
export const bugController = new BugController(bugUseCase);
export const dashboardController = new DashboardController(dashboardUseCase);
export const systemController = new SystemController(db);
